platforms/nuttx/src/px4/stm/stm32_common/spi/spi.cpp

```diff
@@ -0,0 +1,574 @@
+#define CONFIG_STM32_SPI4
+#endif
+#ifdef CONFIG_STM32F7_SPI5
+#define CONFIG_STM32_SPI5
+#endif
+#ifdef CONFIG_STM32F7_SPI6
+#define CONFIG_STM32_SPI6
+#endif
+
+// wrapper for stm32h7
+#ifdef CONFIG_STM32H7_SPI1
+#define CONFIG_STM32_SPI1
+#endif
+#ifdef CONFIG_STM32H7_SPI2
+#define CONFIG_STM32_SPI2
+#endif
+#ifdef CONFIG_STM32H7_SPI3
+#define CONFIG_STM32_SPI3
+#endif
+#ifdef CONFIG_STM32H7_SPI4
+#define CONFIG_STM32_SPI4
+#endif
+#ifdef CONFIG_STM32H7_SPI5
+#define CONFIG_STM32_SPI5
+#endif
+#ifdef CONFIG_STM32H7_SPI6
+#define CONFIG_STM32_SPI6
+#endif
+
+static const px4_spi_bus_t *_spi_bus1;
+static const px4_spi_bus_t *_spi_bus2;
+static const px4_spi_bus_t *_spi_bus3;
+static const px4_spi_bus_t *_spi_bus4;
+static const px4_spi_bus_t *_spi_bus5;
+static const px4_spi_bus_t *_spi_bus6;
+
+static void spi_bus_configgpio_cs(const px4_spi_bus_t *bus)
+{
+	for (int i = 0; i < SPI_BUS_MAX_DEVICES; ++i) {
+		if (bus->devices[i].cs_gpio != 0) {
+			px4_arch_configgpio(bus->devices[i].cs_gpio);
+		}
+	}
+}
+
+__EXPORT void stm32_spiinitialize()
+{
+	px4_set_spi_buses_from_hw_version();
+	board_control_spi_sensors_power_configgpio();
+
+	for (int i = 0; i < SPI_BUS_MAX_BUS_ITEMS; ++i) {
+		switch (px4_spi_buses[i].bus) {
+		case 1: _spi_bus1 = &px4_spi_buses[i]; break;
+
+		case 2: _spi_bus2 = &px4_spi_buses[i]; break;
+
+		case 3: _spi_bus3 = &px4_spi_buses[i]; break;
+
+		case 4: _spi_bus4 = &px4_spi_buses[i]; break;
+
+		case 5: _spi_bus5 = &px4_spi_buses[i]; break;
+
+		case 6: _spi_bus6 = &px4_spi_buses[i]; break;
+		}
+	}
+
+#ifdef CONFIG_STM32_SPI1
+	ASSERT(_spi_bus1);
+
+	if (board_has_bus(BOARD_SPI_BUS, 1)) {
+		spi_bus_configgpio_cs(_spi_bus1);
+	}
+
+#endif // CONFIG_STM32_SPI1
+
+
+#if defined(CONFIG_STM32_SPI2)
+	ASSERT(_spi_bus2);
+
+	if (board_has_bus(BOARD_SPI_BUS, 2)) {
+		spi_bus_configgpio_cs(_spi_bus2);
+	}
+
+#endif // CONFIG_STM32_SPI2
+
+#ifdef CONFIG_STM32_SPI3
+	ASSERT(_spi_bus3);
+
+	if (board_has_bus(BOARD_SPI_BUS, 3)) {
+		spi_bus_configgpio_cs(_spi_bus3);
+	}
+
+#endif // CONFIG_STM32_SPI3
+
+#ifdef CONFIG_STM32_SPI4
+	ASSERT(_spi_bus4);
+
+	if (board_has_bus(BOARD_SPI_BUS, 4)) {
+		spi_bus_configgpio_cs(_spi_bus4);
+	}
+
+#endif // CONFIG_STM32_SPI4
+
+
+#ifdef CONFIG_STM32_SPI5
+	ASSERT(_spi_bus5);
+
+	if (board_has_bus(BOARD_SPI_BUS, 5)) {
+		spi_bus_configgpio_cs(_spi_bus5);
+	}
+
+#endif // CONFIG_STM32_SPI5
+
+
+#ifdef CONFIG_STM32_SPI6
+	ASSERT(_spi_bus6);
+
+	if (board_has_bus(BOARD_SPI_BUS, 6)) {
+		spi_bus_configgpio_cs(_spi_bus6);
+	}
+
+#endif // CONFIG_STM32_SPI6
+}
+
+static inline void stm32_spixselect(const px4_spi_bus_t *bus, struct spi_dev_s *dev, uint32_t devid, bool selected)
+{
+	int matched_dev_idx = -1;
+
+	for (int i = 0; i < SPI_BUS_MAX_DEVICES; ++i) {
+		if (bus->devices[i].cs_gpio == 0) {
+			break;
+		}
+
+		if (devid == bus->devices[i].devid) {
+			matched_dev_idx = i;
+
+		} else {
+			// Making sure the other peripherals are not selected
+			stm32_gpiowrite(bus->devices[i].cs_gpio, 1);
```

Comment:
Did you consider not propagating this?

Reply:
That is an option too. I was considering the opposite: dropping the deselect.

Reply:
@bkueng 

>  I was considering the opposite: dropping the deselect.

Do you mean leaving parts selected until another device is selected?

That might be a bad idea, the CS needs to be de-asserted  to release the bus and re-asserted for some parts to reset state (address counters) etc

Reply:
Yes, but since that's a bad idea, let's look into the other option then.